src/AutoMapper/MapperConfiguration.cs

```diff
@@ -109,39 +109,43 @@ public TypeMap ResolveTypeMap(Type sourceType, Type destinationType)
         public TypeMap ResolveTypeMap(TypePair typePair)
         {
             var typeMap = _typeMapPlanCache.GetOrAdd(typePair, _getTypeMap);
-
-            _typeMapPlanCache.AddOrUpdate(typePair, typeMap, (tp, tm) =>
+            if(Configuration.CreateMissingTypeMaps && typeMap != null && typeMap.MapExpression == null)
             {
-                if (tm != null)
-                    lock (tm)
-                        if (tm.Sealed == false)
-                            tm.Seal(_typeMapRegistry, this);
-                return tm;
-            });
-
+                lock(typeMap)
+                {
+                    typeMap.Seal(_typeMapRegistry, this);
+                }
+            }
             return typeMap;
         }
 
-        private TypeMap GetTypeMap(TypePair pair)
+        private TypeMap GetTypeMap(TypePair initialTypes)
         {
-            foreach (var tp in pair.GetRelatedTypePairs())
+            TypeMap typeMap;
+            foreach(var types in initialTypes.GetRelatedTypePairs())
             {
-                TypeMap typeMap;
-                if (_typeMapPlanCache.TryGetValue(tp, out typeMap))
+                if(_typeMapPlanCache.TryGetValue(types, out typeMap))
+                {
                     return typeMap;
-
-                typeMap = FindClosedGenericTypeMapFor(tp, pair);
-
-                if (typeMap != null)
+                }
+                typeMap = FindTypeMapFor(types);
+                if(typeMap != null)
+                {
                     return typeMap;
-
-                if (CoveredByObjectMap(pair))
-                    return null;
-
-                typeMap = FindConventionTypeMapFor(tp);
-
-                if (typeMap != null)
+                }
+                typeMap = FindClosedGenericTypeMapFor(types, initialTypes);
+                if(typeMap != null)
+                {
                     return typeMap;
+                }
+                if(!CoveredByObjectMap(initialTypes))
+                {
+                    typeMap = FindConventionTypeMapFor(types);
+                    if(typeMap != null)
+                    {
+                        return typeMap;
+                    }
+                }

```

Comment:
btw does this make more sense now? I broke it out instead of a ginormous conditional